File: src/engine/mesh.ts
import { mat4, quat, vec3, type Mat4, type Quat, type Vec3 } from "wgpu-matrix";

/*
pub const QUAD_VERTICES: [Vec3; 4] = [
    Vec3::new(-0.5, -0.5, 0.0),
    Vec3::new(0.5, -0.5, 0.0),
    Vec3::new(0.5, 0.5, 0.0),
    Vec3::new(-0.5, 0.5, 0.0),
];
pub const QUAD_INDICES: [u16; 6] = [0, 1, 2, 2, 3, 0];

pub struct Transform {
    position: Point3,
    rotation: glam::Quat,
    scale: f32,
}

impl Transform {
    pub fn to_matrix(&self) -> Matrix4<f32> {
        Matrix4::from_scale_rotation_translation(
            Vector3::new(self.scale, self.scale, self.scale),
            self.rotation,
            self.position.to_vector(),
        )
    }
}

pub struct Mesh {
    pub transform: Transform,
    pub vertex_buffer: wgpu::Buffer,
    pub index_buffer: wgpu::Buffer,
    pub num_indices: u32,
}
impl Mesh {
    pub fn new_quad(device: &wgpu::Device) -> Self {
        let transform = Transform {
            position: Point3::new(0.0, 0.0, 0.0),
            rotation: glam::Quat::IDENTITY,
            scale: 1.0,
        };
        let vertex_buffer_contents = QUAD_VERTICES
            .iter()
            .map(|&position| shader::VertexInput { position })
            .collect::<Vec<_>>();

        let vertex_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some("Vertex Buffer"),
            contents: bytemuck::cast_slice(&vertex_buffer_contents),
            usage: wgpu::BufferUsages::VERTEX,
        });

        let index_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some("Index Buffer"),
            contents: bytemuck::cast_slice(&QUAD_INDICES),
            usage: wgpu::BufferUsages::INDEX,
        });

        Self {
            transform,
            vertex_buffer,
            index_buffer,
            num_indices: QUAD_INDICES.len() as u32,
        }
    }

    pub fn get_model_matrix(&self) -> Matrix4<f32> {
        self.transform.to_matrix()
    }
}
*/

const QUAD_VERTICES: Vec3[] = [
  vec3.create(-0.5, -0.5, 0.0),
  vec3.create(-0.5, -0.5, 0.0),
  vec3.create(0.5, -0.5, 0.0),
  vec3.create(0.5, 0.5, 0.0),
];
const QUAD_INDICES: Uint16Array = new Uint16Array([0, 1, 2, 2, 3, 0]);

export class Transform {
  constructor(
    public position: Vec3 = vec3.create(),
    public rotation: Quat = quat.create(),
    public scale: number = 1.0
  ) {}

  toMatrix(): Mat4 {
    return mat4.mul(
      mat4.scaling(vec3.create(this.scale, this.scale, this.scale)),
      mat4.mul(mat4.fromQuat(this.rotation), mat4.translation(this.position))
    );
  }
}

File: parametric-renderer-core/renderer-core/src/mesh.rs
use glam::{Vec2, Vec3};
use wgpu::util::DeviceExt;

use crate::shaders::shader;

pub struct Mesh {
    pub vertex_buffer: wgpu::Buffer,
    pub index_buffer: wgpu::Buffer,
    pub num_indices: u32,
}
impl Mesh {
    fn with_contents(
        device: &wgpu::Device,
        vertex_buffer_contents: &[shader::VertexInput],
        index_buffer_contents: &[u16],
    ) -> Self {
        let vertex_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some("Vertex Buffer"),
            contents: bytemuck::cast_slice(vertex_buffer_contents),
            usage: wgpu::BufferUsages::VERTEX,
        });

        let index_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some("Index Buffer"),
            contents: bytemuck::cast_slice(index_buffer_contents),
            usage: wgpu::BufferUsages::INDEX,
        });

        Self {
            vertex_buffer,
            index_buffer,
            num_indices: index_buffer_contents.len() as u32,
        }
    }

    /// Create a new tesselated quad mesh
    pub fn new_tesselated_quad(device: &wgpu::Device, split_count: u32) -> Self {
        let (vertices, indices) = tesselated_quad(split_count);

        Mesh::with_contents(device, &vertices, &indices)
    }

    pub fn cubemap_cube(device: &wgpu::Device, min: Vec3, max: Vec3) -> Self {
        let (vertices, indices) = cubemap_cube(min, max);

        Mesh::with_contents(device, &vertices, &indices)
    }
}

fn tesselated_quad(split_count: u32) -> (Vec<shader::VertexInput>, Vec<u16>) {
    let mut vertices = Vec::new();
    let mut indices = Vec::new();

    let quad_count_one_side = split_count + 1;

    for i in 0..=quad_count_one_side {
        for j in 0..=quad_count_one_side {
            let position = Vec3::new(i as f32, j as f32, 0.0) / (quad_count_one_side as f32);
            let uv = Vec2::new(i as f32, j as f32) / (quad_count_one_side as f32);
            vertices.push(shader::VertexInput { position, uv });
        }
    }

    let vertices_per_row = quad_count_one_side + 1;
    for i in 0..quad_count_one_side {
        for j in 0..quad_count_one_side {
            let i0 = i * vertices_per_row + j;
            let i1 = i0 + 1;
            let i2 = i0 + vertices_per_row;
            let i3 = i2 + 1;

            indices.push(i0 as u16);
            indices.push(i1 as u16);
            indices.push(i2 as u16);

            indices.push(i2 as u16);
            indices.push(i1 as u16);
            indices.push(i3 as u16);
        }
    }
    (vertices, indices)
}

fn cubemap_cube(min: Vec3, max: Vec3) -> (Vec<shader::VertexInput>, Vec<u16>) {
    let vertices = [
        (-1, 1, -1),
        (-1, -1, -1),
        (1, -1, -1),
        (1, -1, -1),
        (1, 1, -1),
        (-1, 1, -1),
        //
        (-1, -1, 1),
        (-1, -1, -1),
        (-1, 1, -1),
        (-1, 1, -1),
        (-1, 1, 1),
        (-1, -1, 1),
        //
        (1, -1, -1),
        (1, -1, 1),
        (1, 1, 1),
        (1, 1, 1),
        (1, 1, -1),
        (1, -1, -1),
        //
        (-1, -1, 1),
        (-1, 1, 1),
        (1, 1, 1),
        (1, 1, 1),
        (1, -1, 1),
        (-1, -1, 1),
        //
        (-1, 1, -1),
        (1, 1, -1),
        (1, 1, 1),
        (1, 1, 1),
        (-1, 1, 1),
        (-1, 1, -1),
        //
        (-1, -1, -1),
        (-1, -1, 1),
        (1, -1, -1),
        (1, -1, -1),
        (-1, -1, 1),
        (1, -1, 1),
    ]
    .into_iter()
    .map(|(x, y, z)| shader::VertexInput {
        position: Vec3::new(
            if x < 0 { min.x } else { max.x },
            if y < 0 { min.y } else { max.y },
            if z < 0 { min.z } else { max.z },
        ),
        uv: Vec2::ZERO,
    })
    .collect::<Vec<_>>();
    let indices = (0u16..(vertices.len() as u16)).collect();
    (vertices, indices)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unit_quad() {
        let (vertices, indices) = tesselated_quad(0);

        assert_eq!(vertices.len(), 4);
        assert_eq!(indices.len(), 6);
        assert_eq!(indices, vec![0, 1, 2, 2, 1, 3]);
        assert_eq!(vertices[0].position, Vec3::new(0.0, 0.0, 0.0));
        assert_eq!(vertices[1].position, Vec3::new(0.0, 1.0, 0.0));
        assert_eq!(vertices[2].position, Vec3::new(1.0, 0.0, 0.0));
        assert_eq!(vertices[3].position, Vec3::new(1.0, 1.0, 0.0));
    }
}
